lib/3scale/backend/metric/collection.rb

```diff
@@ -40,20 +40,29 @@ def parse_usage(raw_usage)
           end
         end
 
+        # Propagates the usage to all the levels of the hierarchy.
+        # For example, in this scenario:
+        # m1 --child_of--> m2 --child_of--> m3
+        # If there's a +1 in m1, this method will set the +1 in the other 2 as
+        # well.
         def process_parents(usage)
-          usage.keys.inject(usage.dup) do |memo, id|
+          usage.inject(usage.dup) do |memo, (id, val)|
             p_id = parent_id(id)
-            if p_id
-              if Usage.is_set? memo[id]
-                memo[p_id] = memo[id]
+            is_set_op = Usage.is_set?(val)
+
+            while p_id
```

Comment:
Not really, because `id` is defined in the outer loop.

Reply:
But this is `p_id`, which you only use inside the loop :?

Reply:
You haven't answered this comment, @davidor

Reply:
I did. It doesn't work because `id` is in the outer loop.

So if we wanted to put the condition in the `while`, we'd need to do something like

```ruby
...
current_metric_id = id
while (current_metric_id = parent_id(current_metric_id))
...
end
...
```

Notice that we'd need to initialize the variable anyway, and whether that makes the code more readable is debatable.

Reply:
You don't need to, it's enough to just name it `id` and avoid the special casing of the first `id` not being a real parent id, since you don't do anything particular with it. There's enough context to not need any distinction.

Reply:
How's that more readable? It shadows the `id` defined in `do |memo, (id, val)|`

Reply:
It does not shadow it, the `id` yielded is never used again at all, just to bootstrap the process. IOW, you can extract the inner loop to a method and you'll have exactly the same code, and more readable.

Reply:
I don't have a strong opinion about this. To me, the 2 solutions are roughly equivalent and it kind of comes to personal preference. I've changed it so we can move on.